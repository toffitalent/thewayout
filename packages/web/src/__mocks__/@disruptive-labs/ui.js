const actual = jest.requireActual('@disruptive-labs/ui');

const showToast = jest.fn();

actual.Toast.show = showToast;
actual.Toast.danger = showToast;
actual.Toast.info = showToast;
actual.Toast.success = showToast;
actual.Toast.warning = showToast;

module.exports = {
  ...actual,
  showToast,
};
