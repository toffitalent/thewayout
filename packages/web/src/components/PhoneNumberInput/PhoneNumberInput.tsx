import { Control, Input, InputProps } from '@disruptive-labs/ui';
import { ChangeEvent } from 'react';
import { Controller } from 'react-hook-form';
import { useFormattedPhoneNumber } from '@app/hooks';

interface PhoneNumberInputProps extends InputProps {
  isFieldsRequired?: boolean;
  control: Control<any, any>;
  setPhoneValue: (value: string) => void;
}

export const PhoneNumberInput = ({
  isFieldsRequired = false,
  control,
  setPhoneValue,
  ...props
}: PhoneNumberInputProps) => {
  const { formatNumber, parsePhoneNumber } = useFormattedPhoneNumber();
  const isFormattedNumber = !!formatNumber && !!parsePhoneNumber;

  const validatePhone = (phone: string) => {
    if (phone.replace(/[^0-9]/g, '').length === 10) {
      const phoneNumber = parsePhoneNumber!(phone, 'US');
      if (phoneNumber) {
        return phoneNumber.isValid() ? undefined : 'Invalid phone number';
      }
    }
    if (phone.length !== 14 && phone.length !== 0) {
      return 'Invalid phone number';
    }
    return undefined;
  };

  return (
    <Controller
      name={props.name || 'phone'}
      rules={{
        ...(isFormattedNumber && { validate: validatePhone }),
        required: isFieldsRequired,
      }}
      control={control}
      render={({ field: { onBlur, value, ref }, fieldState: { error, invalid } }) => (
        <Input
          {...props}
          type="text"
          invalid={invalid}
          onBlur={onBlur}
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            if (isFormattedNumber) {
              formatNumber.reset();
              setPhoneValue(formatNumber.input(e.target.value));
            } else {
              setPhoneValue(e.target.value);
            }
          }}
          ref={ref}
          value={value}
          helperText={error?.message}
          maxLength={isFormattedNumber ? 14 : 10}
        />
      )}
    />
  );
};
