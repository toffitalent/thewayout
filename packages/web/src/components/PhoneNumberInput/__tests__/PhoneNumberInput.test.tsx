import { useForm } from '@disruptive-labs/ui';
import { render, renderHook, waitFor } from '@test';
import { PhoneNumberInput } from '../PhoneNumberInput';

jest.mock('@app/hooks', () => ({
  ...jest.requireActual('@app/hooks'),
  useFormattedPhoneNumber: jest
    .fn()
    .mockImplementationOnce(() => ({
      formatNumber: { reset: jest.fn(), input: (value: string) => value },
      parsePhoneNumber: jest.fn(),
    }))
    .mockImplementationOnce(() => ({
      formatNumber: undefined,
      parsePhoneNumber: undefined,
    })),
}));

describe('PhoneNumberInput', () => {
  test('renders correctly', async () => {
    const { result } = renderHook(() => useForm({ defaultValues: { phone: '' } }));

    const { container, findByText } = render(
      <PhoneNumberInput control={result.current.control} setPhoneValue={jest.fn()} label="Phone" />,
    );

    await waitFor(async () => expect(await findByText('Phone')).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

  test('renders correctly when format number package not load', async () => {
    const { result } = renderHook(() => useForm({ defaultValues: { phone: '' } }));

    const { container, findByText } = render(
      <PhoneNumberInput control={result.current.control} setPhoneValue={jest.fn()} label="Phone" />,
    );

    await waitFor(async () => expect(await findByText('Phone')).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });
});
