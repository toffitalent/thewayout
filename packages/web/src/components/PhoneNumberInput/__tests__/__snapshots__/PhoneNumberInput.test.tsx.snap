// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberInput renders correctly 1`] = `
<div>
  <div
    class="control"
    role="group"
  >
    <label
      class="label"
      for=":r0:"
    >
      Phone
    </label>
    <div
      class="root primary md md"
    >
      <input
        class="input"
        id=":r0:"
        maxlength="14"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput renders correctly when format number package not load 1`] = `
<div>
  <div
    class="control"
    role="group"
  >
    <label
      class="label"
      for=":r1:"
    >
      Phone
    </label>
    <div
      class="root primary md md"
    >
      <input
        class="input"
        id=":r1:"
        maxlength="10"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
